### PR TITLE
[numpy] FFI for insert \ delete \ matmul etc.

### PR DESCRIPTION
**Pull request by JiangZhaoh**
## Description ##
ffi op list:
| op | Old FFI (ctypes) (us) | New FFI (cython) (us) |
| :-: | :-------------------: | :---------------------: |
| append | 31.3 | 7.73 |
| orange | 48.97 | 12.93 |
| blackman | 47.84 | 33.94 |
| concatenate | 34.3 | 8.84 |
| delete | 36.61 | 8.26 |
| eye | 50.8 | 13.13 |
| hamming | 44.78 | 11.95 |
| hanning | 44.69 | 11.82 |
| insert | 31.84 | 8.13 |
| append | 31.3 | 7.73 |
| linspace | 40.8 | 8.68 |
| logspace | 57.78 | 13.34 |
| matmul | 25.11 | 10.61 |
| mean | 33.33 | 21.71 |
| random.gamma | 69.77 | 14.74 |
| random.normal | 67.61 | 15.06 |
| random.uniform | 59.54 | 13.18 |
| append | 31.3 | 7.73 |

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/api/operator/numpy/np_delete_op.cc [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+ */
+#include "../utils.h"
+#include "../../../operator/numpy/np_delete_op-inl.h"
+
+namespace mxnet {
+
+MXNET_REGISTER_API("_npi.delete")
+.set_body([](runtime::MXNetArgs args, runtime::MXNetRetValue* ret) {
+  using namespace runtime;
+  const nnvm::Op* op = Op::Get("_npi_delete");
```
Comment (reminisce) [inline src/api/operator/numpy/np_delete_op.cc]: Add `static`.

File: src/api/operator/numpy/np_delete_op.cc [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+                               args[1].operator mxnet::NDArray*()};
+      if (args[2].type_code() == kNull) {
+        param.axis = dmlc::nullopt;
+      } else {
+        param.axis = args[2].operator int();
+      }
+      num_inputs = 2;
+    }
+  } else {
+    inputs = new NDArray*[1]{args[0].operator mxnet::NDArray*()};
```
Comment (reminisce) [inline src/api/operator/numpy/np_delete_op.cc]: You can use `vector<NDArray*>` for `inputs` and pass `inputs.data()` to the funciton `Invoke`. Using `new` is dangerous if manual `delete` is missing.

File: src/api/operator/numpy/np_insert_op.cc [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+namespace mxnet {
+
+MXNET_REGISTER_API("_npi.insert_scalar")
+.set_body([](runtime::MXNetArgs args, runtime::MXNetRetValue* ret) {
+  using namespace runtime;
+  const nnvm::Op* op = Op::Get("_npi_insert_scalar");
+  nnvm::NodeAttrs attrs;
+  op::NumpyInsertParam param;
+  NDArray** inputs = NULL;
```
Comment (reminisce) [inline src/api/operator/numpy/np_insert_op.cc]: Same here.

File: src/api/operator/numpy/np_insert_op.cc [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+  *ret = ndoutputs[0];
+});
+
+MXNET_REGISTER_API("_npi.insert_slice")
+.set_body([](runtime::MXNetArgs args, runtime::MXNetRetValue* ret) {
+  using namespace runtime;
+  const nnvm::Op* op = Op::Get("_npi_insert_slice");
+  nnvm::NodeAttrs attrs;
+  op::NumpyInsertParam param;
+  NDArray** inputs = NULL;
```
Comment (reminisce) [inline src/api/operator/numpy/np_insert_op.cc]: Same here.

File: src/api/operator/numpy/np_insert_op.cc [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+MXNET_REGISTER_API("_npi.insert_tensor")
+.set_body([](runtime::MXNetArgs args, runtime::MXNetRetValue* ret) {
+  using namespace runtime;
+  const nnvm::Op* op = Op::Get("_npi_insert_tensor");
+  nnvm::NodeAttrs attrs;
+  op::NumpyInsertParam param;
+  param.start = dmlc::nullopt;
+  param.step = dmlc::nullopt;
+  param.stop = dmlc::nullopt;
+  NDArray** inputs = NULL;
```
Comment (reminisce) [inline src/api/operator/numpy/np_insert_op.cc]: Same here.

File: src/operator/tensor/init_op.h [hunk 6]
```diff
@@ -215,6 +215,21 @@ struct RangeParam : public dmlc::Parameter<RangeParam> {
     MXNET_ADD_ALL_TYPES
     .describe("Target data type.");
   }
+  void SetAttrDict(std::unordered_map<std::string, std::string>* dict) {
```
Comment (haojin2) [inline src/operator/tensor/init_op.h]: one more blank line above.

File: src/operator/tensor/init_op.h [hunk 7]
```diff
@@ -215,6 +215,21 @@ struct RangeParam : public dmlc::Parameter<RangeParam> {
...
+    step_s << step;
+    repeat_s << repeat;
+    infer_range_s << infer_range;
+    dtype_s << dtype;
+    (*dict)["start"] = start_s.str();
+    (*dict)["stop"] = stop_s.str();
+    (*dict)["step"] = step_s.str();
+    (*dict)["repeat"] = repeat_s.str();
+    (*dict)["infer_range"] = infer_range_s.str();
+    (*dict)["dtype"] = dtype_s.str();
```
Comment (haojin2) [inline src/operator/tensor/init_op.h]: @hzfan shall we use `String2MXNetType` function here?

File: src/operator/numpy/np_window_op.h [hunk 8]
```diff
@@ -58,6 +58,13 @@ struct NumpyWindowsParam : public dmlc::Parameter<NumpyWindowsParam> {
     MXNET_ADD_ALL_TYPES
     .describe("Data-type of the returned array.");
   }
+  void SetAttrDict(std::unordered_map<std::string, std::string>* dict) {
+    std::ostringstream M_s, dtype_s;
+    M_s << M;
+    dtype_s << dtype;
```
Comment (hzfan) [inline src/operator/numpy/np_window_op.h]: Please use `String2MXNetTypeWithBool`  in `op_utils.h` here.
```
(*dict)["dtype"] = String2MXNetTypeWithBool(dtype);
```

File: src/operator/numpy/np_init_op.h [hunk 9]
```diff
@@ -62,6 +62,17 @@ struct NumpyEyeParam : public dmlc::Parameter<NumpyEyeParam> {
...
+  void SetAttrDict(std::unordered_map<std::string, std::string>* dict) {
+    std::ostringstream N_s, M_s, k_s, dtype_s;
+    N_s << N;
+    M_s << M;
+    k_s << k;
+    dtype_s << dtype;
+    (*dict)["N"] = N_s.str();
+    (*dict)["M"] = M_s.str();
+    (*dict)["k"] = k_s.str();
+    (*dict)["dtype"] = dtype_s.str();
```
Comment (hzfan) [inline src/operator/numpy/np_init_op.h]: Please use `String2MXNetTypeWithBool` in `op_utils.h`

File: src/api/operator/numpy/np_window_op.cc [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+ */
+#include <mxnet/api_registry.h>
+#include <mxnet/runtime/packed_func.h>
+#include "../utils.h"
+#include "../../../operator/numpy/np_window_op.h"
+
+namespace mxnet {
+
+inline static void SetNumpyWindowsParam(runtime::MXNetArgs args,
+                                 runtime::MXNetRetValue* ret,
```
Comment (hzfan) [inline src/api/operator/numpy/np_window_op.cc]: Please align

File: src/api/operator/numpy/np_matmul_op.cc [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+  nnvm::NodeAttrs attrs;
+  int num_inputs = 2;
+  NDArray* inputs[2] = {args[0].operator mxnet::NDArray*(),
+                        args[1].operator mxnet::NDArray*()};
+  attrs.op = op;
+  NDArray* out = args[2].operator mxnet::NDArray*();
+  NDArray** outputs = out == nullptr ? nullptr : &out;
+  int num_outputs = out != nullptr;
+  auto ndoutputs = Invoke(op, &attrs, num_inputs, inputs, &num_outputs, outputs);
+  *ret = ndoutputs[0];
```
Comment (hzfan) [inline src/api/operator/numpy/np_matmul_op.cc]: Please use `PythonArg` for `out`

**Comment by JiangZhaoh** [comment]
@mxnet-bot run ci [windows-gpu]

**Comment by JiangZhaoh** [comment]
@mxnet-bot run ci [ci/jenkins/mxnet-validation/windows-gpu]

**Comment by JiangZhaoh** [comment]
@mxnet-bot ci/jenkins/mxnet-validation/windows-gpu

**Comment by JiangZhaoh** [comment]
@mxnet-bot run ci [windows-gpu]

**Comment by haojin2** [comment]
@mxnet-bot run ci [windows-gpu]

**Comment by JiangZhaoh** [comment]
@mxnet-bot run ci [windows-gpu]

**Comment by hzfan** [comment]
LGTM

**Comment by JiangZhaoh** [comment]
run ci [centos-cpu]

**Comment by JiangZhaoh** [comment]
@mxnet-bot run ci [unix-gpu]

**Comment by JiangZhaoh** [comment]
@mxnet-bot run ci [unix-gpu]

**Comment by JiangZhaoh** [comment]
@mxnet-bot run ci [unix-gpu]

**Comment by JiangZhaoh** [comment]
@mxnet-bot run ci [unix-gpu]

**Comment by JiangZhaoh** [comment]
@mxnet-bot run ci [unix-gpu]

**Pull request merged**